19.0 code review

vertiond opened the pull request:
Vertcoin-Core [19.0-rc1 is a rebase](https://github.com/vertcoin-project/vertcoin-core/tree/19.0) on upstream Bitcoin-Core [22.x](https://github.com/bitcoin/bitcoin/tree/22.x)

LarryRuane committed:
Close minor startup race between main and scheduler threads

vtc: fixes startup issue relating to verthash generation

Don't schedule class PeerManagerImpl's background tasks from its
constructor, but instead do that from a separate method,
StartScheduledTasks(), that can be called later at the end of startup,
after other things, such as the active chain, are initialzed.

vertiond committed:
vtc: chainparams

vertiond committed:
vtc: feerates, max money, versionbits

vertiond committed:
vtc: validation

 Add GetPowHash, nHeight, IsWitnessEnabled

 CheckBlockHeader needs g_chainman for LookupBlockIndex

vertiond committed:
vtc: use GMP for kimoto gravity well arithmetic

From PR #114 commit https://github.com/vertcoin-project/vertcoin-core/commit/79e8a0b227e842ba604b2b4f64107ed6f3c0fb2e

vertiond committed:
vtc: Kimoto Gravity Well (KGW) difficulty-readjustment algorithm

vertiond committed:
vtc: Scrypt-N, Lyra2RE, Lyra2REv2, Lyra2REv3, Verthash PoW

Verthash implementation: PR #112 commit
https://github.com/vertcoin-project/vertcoin-core/commit/ae9c9b6bac70fe2d53e032ecec1cf6efa9673219

vertiond committed:
vtc: startup verification

Relevant PRs: #31, #33, #184, #187

vertiond committed:
vtc: ban nodes that keep requesting the same headers

From PR #103

vertiond committed:
vtc: adjust ports

vertiond committed:
vtc: add copyright string for Vertcoin developers

vertiond committed:
vtc: guix builds using -fPIC, --disable-shared

See PR #174

vertiond committed:
vtc: logo branding

Update contrib/debian/copyright

vertiond committed:
vtc: qt, rpc branding

vertiond committed:
vtc: src and executable branding

vertiond committed:
vtc: contrib, manpages branding

vertiond committed:
vtc: Windows code signing certificate

See PR #167

vertiond committed:
vtc: update seeds from jk14's seeder and skip filtering by ASN

ASN filters out up to half of the nodes while VTC does not have many to begin with
See PR #175

vertiond committed:
vtc: locale branding

vertiond committed:
vtc: README files

vertiond committed:
build: bump version to 19.0rc1

vertiond committed:
builder keys: add vertion (#195)

cruelnovo committed:
Update keys.txt with fingerprint for cruelnovo (#196)

Add PGP Fingerprint for Matthew Cummings

KforG committed:
add builder key for KforG (#197)

sylvesternumski committed:
Add key for numski (#198)

cruelnovo reviewed (commented):
## In prior release but not in new release (non-critical) ##
contrib/devtools/gen-manpages.sh # Can we generate Vertcoin branded manpages?

contrib/seeds/makeseeds.py # Slight differences between v0.18 & v19

src/test/netbase_tests.cpp  # Looks like the ports were not updated for Vertcoin

src/qt/bitcoinstrings.cpp  # `*vertcoin_strings[]` vs `*bittcoin_strings[]` - 0.18 has it as `*bittcoin_strings[]` - also could use a QT_TRANSLATE_NOOP("vertcoin-core", "Vertcoin Core") in the file

src/Makefile.am # How come bignum.h was removed?  Line 700, (vertcoin_bin_ldadd) should be used as it was in v0.18?

## Vertcoin rebranding strings ##
doc/man/ # Since we're renaming all the manpage files to 'vertcoin-X.1', should we also update the contents to be vertcoin branded?

src/qt/guiutil.cpp # Vertcoin Core 0.18 changed bitcoin to vertcoin in comments

src/qt/intro.cpp # Vertcoin Core 0.18 changed bitcoin to vertcoin in comments

src/qt/sendcoinsdialog.cpp  # Missed some Vertcoin branding and PSVT vs PSBT crops up again

src/qt/forms/receiverequestdialog.ui # Missed an update for vtc1 and VTC string

src/qt/bitcoingui.cpp  # PSBT vs PSVT - 0.18 has it PSVT (Partially Sent Vertcoin Transaction)

src/rpc/misc.cpp # Vertcoin Core 0.18 changed bitcoin to vertcoin in comments

src/wallet/test/wallet_tests.cpp  # Update strings from bitcoin to vertcoin

src/bitcoin-cli.cpp # Update for a number of strings from bitcoin to vertcoin

src/bitcoind.cpp # Update for a number of strings from bitcoin to vertcoin

## Vertcoin not updated in comments (but was updated in prior release) ##
src/util/system.cpp # Vertcoin Core 0.18 changed bitcoin to vertcoin in comments.  How is the CopyrightHolders function working? Why the strPrefix2?

src/policy/fees.cpp # only a difference in a comment. BTC-per-kb vs VTC-per-kb

src/Makefile.qt.include # Vertcoin Core 0.18 changed bitcoin to vertcoin in comments

## General questions ##
src/util/system.h # How is the CopyrightHolders function working? Why the strPrefix2?

src/crypto/verthash.cpp # How did you know to use `gArgs.GetDataDirNet()` instead of `GetDataDir()` (0.18 uses `GetDataDir()`)

src/crypto/verthash_datafile.cpp # How did you know to use `gArgs.GetDataDirNet()` instead of `GetDataDir()` (0.18 uses `GetDataDir()`)

contrib/seeds/nodes_test.txt  # can we provide a list of known testnet seeds? or remove this file?

contrib/verifybinaries/ # Should we remove this directory from Vertcoin?

src/bignum.h # is this part of KimotoGravityWell?  It is used by pow.cpp

src/chainparamsseeds.h # How come the seeds are not needed for Vertcoin in BIP155 format?

src/pow.cpp # How did you know to fix the "blockstogoback" logic (line 87-94)

configure.ac  # how come we're no longer enabling the SSE2 compiler extension?

src/chainparams.cpp # Should we set nPowTargetTimespan to 3.5 days for SigNet and RegTest?

src/chainparams.cpp # Should we set 2.5 min nPowTargetSpacing for SigNet adn RegTest?

## Scheduler related questions ##
src/test/util/setup_common.cpp # Why was *m_node.scheduler, removed from line 202?

src/test/denialofservice_tests.cpp # Why was *m_node.scheduler removed from peerman call?

src/init.cpp # Why was *m_node.scheduler removed from peerman call? Explain to me how the node peer scheduler works?

cruelnovo commented on src/policy/fees.cpp:
```diff
@@ -562,7 +562,7 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
     trackedTxs++;
 
-    // Feerates are stored and reported as BTC-per-kb:
+    // Feerates are stored and reported as VTC-per-kb:
```
Line 595: update comment to VTC-per-kb

vertiond replied:
I think we can leave comments and anything that is not user facing.

cruelnovo commented on src/Makefile.am:
```diff
@@ -657,91 +685,91 @@ bitcoin_bin_ldadd = \
+vertcoin_node_CPPFLAGS = $(vertcoin_bin_cppflags)
+vertcoin_node_CXXFLAGS = $(vertcoin_bin_cxxflags)
+vertcoin_node_LDFLAGS = $(vertcoin_bin_ldflags)
+vertcoin_node_LDADD = $(LIBBITCOIN_SERVER) $(bitcoin_bin_ldadd) $(LIBBITCOIN_IPC) $(LIBMULTIPROCESS_LIBS)
```
```suggestion
vertcoin_node_LDADD = $(LIBBITCOIN_SERVER) $(vertcoin_bin_ldadd) $(LIBBITCOIN_IPC) $(LIBMULTIPROCESS_LIBS)
```

cruelnovo commented on src/qt/bitcoinstrings.cpp:
```diff
@@ -8,234 +8,234 @@
-QT_TRANSLATE_NOOP("bitcoin-core", "The %s developers"),
-QT_TRANSLATE_NOOP("bitcoin-core", ""
-"%s corrupt. Try using the wallet tool bitcoin-wallet to salvage or restoring "
+static const char UNUSED *vertcoin_strings[] = {
```
```suggestion
static const char UNUSED *bitcoin_strings[] = {
```

cruelnovo commented on src/qt/bitcoinstrings.cpp:
```diff
@@ -8,234 +8,234 @@
-QT_TRANSLATE_NOOP("bitcoin-core", ""
-"%s corrupt. Try using the wallet tool bitcoin-wallet to salvage or restoring "
+static const char UNUSED *vertcoin_strings[] = {
+QT_TRANSLATE_NOOP("vertcoin-core", "The %s developers"),
```
```suggestion
QT_TRANSLATE_NOOP("vertcoin-core", "Vertcoin Core"),
QT_TRANSLATE_NOOP("vertcoin-core", "The %s developers"),
```

vertiond replied:
What is the purpose of this?

vertiond replied:
Seems that bitcoinstrings.cpp needs no modifying for vertcoin and is only necessary for the translation process of which we have left entirely to upstream.

cruelnovo commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -65,7 +65,7 @@
+      <string notr="true">vertcoin:VTC1…</string>
      </property>
      <property name="textFormat">
       <enum>Qt::RichText</enum>
```
At Line 129: replace BTC with VTC

sylvesternumski replied:
Not sure if it matters here, but you replaced BC with VTC.

vertiond replied:
Yeah, our bech32 address prefix is VTC whereas upstream's is BC

cruelnovo reviewed (commented):
A few of my review change notes are remaining on the PR, I could not comment on an appropriate line (line not part of the commit).

cruelnovo commented on src/qt/forms/receiverequestdialog.ui:
```diff
@@ -65,7 +65,7 @@
+      <string notr="true">vertcoin:VTC1…</string>
      </property>
      <property name="textFormat">
       <enum>Qt::RichText</enum>
```
Line 100: <string notr="true">bc1…</string>
Update to:
<string notr="true">vtc1…</string>

vertiond committed:
vtc review: missed branding items (#207)

vertiond committed:
backports from upstream 22.1 (#208)

* rpc: Capture potentially large UniValue by ref for rpcdoccheck

Github-Pull: 25237
Rebased-From: 20ff4991e548630d7bb5e491fa4d69ec49369872

* Add pure Python RIPEMD-160

Github-Pull: 23716
Rebased-From: ad3e9e1f214d739e098c6ebbd300da5df1026a44

* Swap out hashlib.ripemd160 for own implementation

Github-Pull: 23716
Rebased-From: 5b559dc7ecf37ab1604b75ec8ffe8436377a5fb1

* windeploy: Renewed windows code signing certificate

Github-Pull: #25201
Rebased-From: 7e9fe6d800ee8f3381e8f6ad2371f7775c68fad9

* Prevent data race for `pathHandlers`

Github-Pull: bitcoin/bitcoin#25983
Rebased-From: 4296dde28757d88a7076847484669fb202b47bc8

* Revert "build: Use Homebrew's sqlite package if it is available"

This reverts ee7b84e63cbeadd5e680d69ff0548275581e9241 from #20527.
This change was made without any rationale, maybe other than a brew
installed version might be newer, and that's "better". However when
building from source on macOS, it just results in drastically worse
perofrmance, and results in issues / confusions like #25724.

Resolves the "build from source" portion of #25724. Building from
depends is still not ideal, however I have some other changes that might
help improve things in that case.

The difference in performance can be observed using the example from
https://github.com/bitcoin/bitcoin/issues/25724#issuecomment-1213554922,
but minified to only 10 descriptors. i.e:
```bash
time src/bitcoin-cli createwallet speedy true
time src/bitcoin-cli importdescriptors '[
  {"desc":"raw(00145846369f3d6ba366d6f5a903fb5cf4dca3763c0e)#k9wh6v62","timestamp":"now"},
  {"desc":"raw(001420800aabf13f3a4c4ce3ce4c66cecf1d17f21a6e)#6m0hlfh4","timestamp":"now"},
  {"desc":"raw(0014c6bf9715e06d73ebf9b3b02d5cc48d24d8bbabc1)#wyavh36r","timestamp":"now"},
  {"desc":"raw(00141ba7807b3f46af113beaea5c698428ce7138cd8a)#jctdsups","timestamp":"now"},
  {"desc":"raw(00140c1bd27f10fff01b36ddf3c1febaa1acff19b080)#9s6nc3pk","timestamp":"now"},
  {"desc":"raw(00141226e31987e4bc2e63c0ee12908f675e40464b20)#9pp7qm39","timestamp":"now"},
  {"desc":"raw(0014f73f149f7503960a5e849c6ee7a8a8c336f631cb)#qtkxv9fc","timestamp":"now"},
  {"desc":"raw(0014c8ccb4d81ffc769fc5fdd8d7eed69b0e0cae5749)#hn39qayv","timestamp":"now"},
  {"desc":"raw(001498565aead2d67a22a6021d55210f2a917fc22169)#6ar3vwsx","timestamp":"now"},
  {"desc":"raw(001403013248ac0cd9eabe176cad162cda2a19f771e1)#4m47mukd","timestamp":"now"}
]'
```

Running master, when building from souce and using brew installed
sqlite, this takes ~3.4s. With this PR, the same operation takes ~0.3s.

Github-Pull: #25985
Rebased-From: d216d714aae36e6f1c95f82aef81a0be74dee2f3

* doc: remove brew install sqlite from macOS docs

* Adjust `.tx/config` for new Transifex CLI

The old Transifex Command-Line Tool is considered deprecated (as of
January 2022) and will sunset on Nov 30, 2022.

See: https://github.com/transifex/cli/blob/devel/README.md#migrating-from-older-versions-of-the-client

An accompanying PR: https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/142

Github-Pull: #26321
Rebased-From: d6adbb7ee1de661ad89879609eecd11129322405

* rpc: fix crash in deriveaddresses when derivation index is 2147483647

2147483647 is the maximum positive value of a signed int32, and - currently -
the maximum value that the deriveaddresses bitcoin RPC call accepts as
derivation index due to its input validation routines.

Before this change, when the derivation index (and thus range_end) reached
std::numeric_limits<int_32_t>::max(), the "i" variable in the for cycle (which
is declared as int, and as such 32 bits in size on most platforms) would be
incremented at the end of the first iteration and then warp back to
-2147483648. This caused SIGABRT in bitcoind and a core dump.

This change assigns "i" an explicit size of 64 bits on every platform,
sidestepping the problem.

Fixes #26274.

Github-Pull: #26275
Rebased-From: addf9d6502db12cebcc5976df3111cac1a369b82

* rpc: add non-regression test about deriveaddresses crash when index is 2147483647

This test would cause a crash in bitcoind (see #26274) if the fix given in the
previous commit was not applied.

Github-Pull: #26275
Rebased-From: 9153ff3e274953ea0d92d53ddab4c72deeace1b1

* build: Bump version to 22.1.0rc1

* doc: Update manual pages for 22.1.0rc1

* doc: update version number in bips.md to v22.1

* qt: 22.1rc1 translations update

* vtc review: missed branding items

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>
Co-authored-by: Pieter Wuille <pieter@wuille.net>
Co-authored-by: laanwj <126646+laanwj@users.noreply.github.com>
Co-authored-by: MacroFake <falke.marco@gmail.com>
Co-authored-by: Andrew Chow <achow101-github@achow101.com>
Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>
Co-authored-by: fanquake <fanquake@gmail.com>
Co-authored-by: muxator <antonio.muci@bancaditalia.it>

cruelnovo committed:
verthash: modernize boost calls to support future boost releases (#210)

cruelnovo reviewed (approved):
Backports from upstream 22.1 are good

cruelnovo reviewed (approved):
End-user branding for Vertcoin Core is good.  Linking flags for 'vertcoin_node_LDADD' are set.
